### PR TITLE
[processor/transform] Add new delete functions

### LINKED ISSUE #11823: [processor/transform] Add ability to delete entries from a map.
**Is your feature request related to a problem? Please describe.**
The transform processor has no way to delete entries from a map.  If an attribute needs to be delete the only solution right now is `keep_keys`, which is unrealistic.

**Describe the solution you'd like**
Add a `delete_key` and `delete_all_matching_keys` functions.

**Additional context**
[Deleting capability is outlined in the processing doc.](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/processing.md#examples)

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
Adds 2 new functions that allow deleting keys from maps.

**Link to tracking Issue:** 
Closes #11823

**Testing:**
Added new unit tests and integration tests

**Documentation:** 
Updated readme

### CODE REVIEW

File: processor/transformprocessor/README.md [hunk 1]
```diff
@@ -33,6 +33,10 @@ Supported functions:
 e.g., `set(attributes["http.path"], "/foo")`, `set(name, attributes["http.route"])`, `set(trace_state["svc"], "example")`, `set(attributes["source"], trace_state["source"])`. If `value` resolves to `nil`, e.g.
 it references an unset map value, there will be no action.
 
+- `delete_key(target, key)` - `target` is a path expression to a map type field. `key` is a string that is a key in the map.  The key will be deleted from the map.  e.g., `delete_key(attributes, "http.request.header.authorization")`
+
+- `delete_matching_keys(target, pattern)` - `target` is a path expression to a map type field. `pattern` is a regex string.  All keys that match the pattern will be deleted from the map.  e.g., `delete_matching_keys(attributes, ".*\.header\.authorization")`
```
Comment (codeboten) [inline processor/transformprocessor/README.md:38]: Should this be `delete_all_matches` to be consistent with `replace_all_matches`?

Comment (TylerHelmuth) [reply processor/transformprocessor/README.md:38]: I thought about that but I didn't want anyone to think it deleted the matching sections in the map's value vs replacing it like `replace_all_matches`.  I wanted it to be clear that its the whole entry being deleted, so I kept it closer to `delete_key`

**Comment by TylerHelmuth** [comment]
@open-telemetry/collector-contrib-approvers please review

**Comment by anuraaga** [comment]
I think we added `nil` to the language recently. Is there a difference between set nil and delete?

**Comment by TylerHelmuth** [comment]
> I think we added `nil` to the language recently. Is there a difference between set nil and delete?

Wouldn't `set(attribute["test"], nil)` keep that attribute key in the map but set its value to nil?  The delete functions would completely remove the key and value from the map.

**Comment by TylerHelmuth** [comment]
Opened https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/12193 for the set nil issue

**Review by djaglowski — APPROVED** [review]

**Review by anuraaga — APPROVED** [review]

**Pull request merged**